community/index/src/main/java/org/neo4j/index/gbptree/GenSafePointerPair.java

```diff
@@ -134,77 +170,52 @@ public static long read( PageCursor cursor, long stableGeneration, long unstable
+            return buildSuccessfulReadResult( FLAG_SLOT_A, logicalPos, gsppOffset, pointerA );
         }
         else if ( pointerStateB == STABLE )
         {
-            return SUCCESS | READ | pointerB;
+            return buildSuccessfulReadResult( FLAG_SLOT_B, logicalPos, gsppOffset, pointerB );
         }
 
-        return FAIL | READ | generationState( generationA, generationB ) |
-               ((long) pointerStateA) << STATE_SHIFT_A | ((long) pointerStateB) << STATE_SHIFT_B;
-
-//        TODO ANOTHER APPROACH, keep here for reference and pursuit later
-//        boolean aIsReadable = correctChecksumA && generationA >= MIN_GENERATION &&
-//                (generationA <= stableGeneration || generationA == unstableGeneration);
-//        boolean bIsReadable = correctChecksumB && generationB >= MIN_GENERATION &&
-//                (generationB <= stableGeneration || generationB == unstableGeneration);
-//        // Success cases
-//        // - both valid and different generation
-//        if ( aIsReadable && bIsReadable && generationA != generationB )
-//        {
-//            return generationA > generationB ? pointerA : pointerB;
-//        }
-//        // - one valid
-//        else if ( aIsReadable ^ bIsReadable )
-//        {
-//            return aIsReadable ? pointerA : pointerB;
-//        }
-//        else
-//        {
-//            if ( isEmpty( generationA, pointerA ) && isEmpty( generationB, pointerB ) )
-//            {
-//                return POINTER_UNDECIDED_BOTH_EMPTY;
-//            }
-//            else if ( !aIsReadable && !bIsReadable && generationA != generationB )
-//            {
-//                return POINTER_UNDECIDED_NONE_CORRECT;
-//            }
-//            else if ( generationA == generationB )
-//            {
-//                return POINTER_UNDECIDED_SAME_GENERATION;
-//            }
-//            else
-//            {
-//                throw new UnsupportedOperationException( "Uncovered case" +
-//                                                         " A: " + generationA + "," + pointerA + "," + checksumA +
-//                                                         " B: " + generationB + "," + pointerB + "," + checksumB );
-//            }
-//        }
+        return FLAG_FAIL | FLAG_READ | generationState( generationA, generationB ) |
+               ((long) pointerStateA) << SHIFT_STATE_A | ((long) pointerStateB) << SHIFT_STATE_B;
+    }
+
+    private static long buildSuccessfulReadResult( long slot, int logicalPos, int gsppOffset, long pointer )
+    {
+        boolean isLogicalPos = logicalPos != NO_LOGICAL_POS;
+        long offsetType = isLogicalPos ? FLAG_LOGICAL_POS : FLAG_ABS_OFFSET;
+        long genOffset = isLogicalPos ? logicalPos : gsppOffset;
+        if ( (genOffset & ~MAX_GEN_OFFSET_MASK) != 0 )
+        {
+            throw new IllegalArgumentException( "Illegal genOffset:" + genOffset + ", it would be too large, max is " +
```

Comment:
hmm, actually, we _are_ safe here because keyCount check above guards for that, `KeySearch.search` cannot go outside 0 and keyCount. SImply extended the keyCount check for negative numbers as well.